query-service-impl/src/main/java/org/hypertrace/core/query/service/AttributeClientProvider.java

```diff
@@ -0,0 +1,30 @@
+package org.hypertrace.core.query.service;
+
+import static org.hypertrace.core.grpcutils.client.RequestContextClientCallCredsProviderFactory.getClientCallCredsProvider;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import org.hypertrace.core.attribute.service.cachingclient.CachingAttributeClient;
+
+class AttributeClientProvider implements Provider<CachingAttributeClient> {
```

Comment:
nit, can/should this be final?

Reply:
No reason - I can't imagine a scenario where this would be extended. Unlike variables, I don't generally mark classes as final by default, but I don't really have any justification for it since the same reasoning for variables applies to classes, if not more so (preventing the inheritance scourge!).

Thanks for pointing this out, will update these two spots and try to get into the habit of doing this going forward.

Reply:
Thinking this through a bit more - I guess, as you had already pointed out, this really applies for implementations that shouldn't be referenced directly. A big reason not to is it makes things difficult to mock, but in these scenarios (and any where the implementation shouldn't be referenced directly) the interface gives you an adequate and more appropriate handle to mock anyway.

Reply:
I was gonna say final classes cannot be mocked but @aaron-steinfeld brought up a good point about using interfaces instead of direct impls.